### PR TITLE
ENH: Add locking to umath_linalg if no lapack is detected at build time

### LINKED ISSUE #22509: BUG: Lapack lite is not thread-safe (need to guard)
`skimage` found an issue probably related to `eigvalsh` or generally threaded calls to `linalg_umath.cpp`.

Unfortunately, I have not yet narrowed it down further, and the test is slightly flaky (on my machine).  For example, it seems to easier reproduce running the full file, rather than just the test.

To reproduce, I have to explicitly use **`NPY_BLAS_ORDER=`** (disable blas use), then run current main of skimage tests:
```
python3 -m pytest skimage/feature/tests/test_basic_features.py::test_multiscale_basic_features_deprecated_multichannel
```

I have a gdb traceback of a crash (the actual crash fluctuates a bit, so it may well be elsewhere):
```
#5  0x00007fffb7ef3f7f in release_evd<double> (params=0x7fffda26e940) at numpy/linalg/umath_linalg.cpp:1482
#6  eigh_wrapper<double> (JOBZ=<optimized out>, UPLO=<optimized out>, args=<optimized out>, dimensions=<optimized out>, 
    steps=<optimized out>) at numpy/linalg/umath_linalg.cpp:1555
#7  0x00007ffff639eed5 in generic_wrapped_legacy_loop (__NPY_UNUSED_TAGGEDcontext=<optimized out>, data=<optimized out>, 
    dimensions=<optimized out>, strides=<optimized out>, auxdata=0x7fffd19e9b30)
    at numpy/core/src/umath/legacy_array_method.c:87
```

So it seems eigvalsh and threading related.  I need to dig further eventually, but maybe someone else has an idea what to look into (the umath code, the blas/lapack fallback code?).

### PR DESCRIPTION
**Pull request by ngoldbaum**
Closes #22509

This is a re-do of #26687 following the suggestion in https://github.com/numpy/numpy/pull/26687#issuecomment-2167777366.

This adds locking macros around all the calls into the low-level LAPACK code in umath_linalg.cpp. The macros are either no-ops or calls to lock or unlock a global mutex declared statically in the `_umath_linalg` module and initialized during that modules' initialization. The mutex is only used if we do not detect a LAPACK during the build, indicating that lapack_lite is being used.

The test I added randomly fails or segfaults on `main` if you force numpy to build using lapack_lite.

Happy to add more extensive thread safety testing if there's a desire for it.

I think I got all the low-level calls but please let me know if you think I missed something.

### CODE REVIEW

**Comment by ngoldbaum** [comment]
This needs a release note I think since lapack_lite has been thread-unsafe since forever. Also all the work I’m doing toward improving and documenting thread safety probably needs a release note. Just commenting to remind myself to open a followup issue about a release note for free-threaded support and a release note specific to this, since this thread safety issue is a problem even with the GIL.

File: numpy/_core/config.h.in [hunk 1]
```diff
@@ -108,6 +108,8 @@
 #mesondefine HAVE_LDOUBLE_IBM_DOUBLE_DOUBLE_LE
 #mesondefine HAVE_LDOUBLE_IBM_DOUBLE_DOUBLE_BE
 
+#mesondefine HAVE_BLAS
```
Comment (seberg) [inline numpy/_core/config.h.in:111]: Style question (I really don't know if there is a real preference).  Should we use `NPY_` prefix for new additions?

File: numpy/_core/meson.build [hunk 2]
```diff
@@ -506,6 +506,10 @@ if cc.has_function_attribute('visibility:hidden') and host_machine.system() != '
 endif
 cdata.set('NPY_VISIBILITY_HIDDEN', visibility_hidden)
 
+# if set, we're using lapack_lite
+if have_blas
+  cdata.set10('HAVE_BLAS', have_blas)
+endif
```
Comment (seberg) [inline numpy/_core/meson.build:512]: Pretty sure we can have blas, but not lapack, so I think this should use `have_lapack`.
(Not sure this is actually plausible/possible in practice, but seems more specific anyway.)

**Review by seberg — COMMENTED** [review]
Nice straight forward!  I happy to not test every function here, that might be nice but rather repetitive anyway.
One small comment mostly, I think we should use `have_lapack` (even if it  may not matter in practice)?

Was considering if there are other blas/lapacks that are not thread-safe.  But I think renaming the lock is a bridge we can cross when that happens.

File: numpy/_core/config.h.in (hunk 1 above)
Comment (ngoldbaum) [reply numpy/_core/config.h.in:111]: The `NPY` ones go in the public `numpyconfig.h` header. This is only exposed in the private `config.h` header (included by `npy_config.h`, another internal header).

File: numpy/_core/meson.build (hunk 2 above)
Comment (rgommers) [reply numpy/_core/meson.build:512]: It's very rare, but possible indeed. I think older versions of Arch Linux shipping a BLAS-only OpenBLAS build is the one case that I saw more than once - but that is gone now.

Comment (rgommers) [reply numpy/_core/meson.build:512]: Maybe a little more descriptive would be nice: `HAVE_EXTERNAL_LAPACK`.

File: numpy/_core/meson.build [hunk 3]
```diff
@@ -506,6 +506,10 @@ if cc.has_function_attribute('visibility:hidden') and host_machine.system() != '
 endif
 cdata.set('NPY_VISIBILITY_HIDDEN', visibility_hidden)
 
+# if set, we're using lapack_lite
```
Comment (rgommers) [inline numpy/_core/meson.build:509]: typo: we're then _not_ using `lapack_lite`

**Review by rgommers — COMMENTED** [review]
LGTM modulo two quite minor comments. Thanks @ngoldbaum

File: numpy/_core/meson.build (hunk 2 above)
Comment (ngoldbaum) [reply numpy/_core/meson.build:512]: Thanks for pointing this out, I conflated with the `allow-noblas` build option.

File: numpy/_core/meson.build [hunk 4]
```diff
@@ -506,6 +506,10 @@ if cc.has_function_attribute('visibility:hidden') and host_machine.system() != '
 endif
 cdata.set('NPY_VISIBILITY_HIDDEN', visibility_hidden)
 
+# if not set, we're using lapack_lite
+if have_lapack
+  cdata.set10('HAVE_EXTERNAL_LAPACK', have_blas)
```
Comment (rgommers) [inline numpy/_core/meson.build:511]: Should be `have_lapack` here, not `have_blas`.

File: doc/release/upcoming_changes/26750.improvement.rst [hunk 5]
```diff
@@ -0,0 +1,12 @@
+`lapack_lite` is thread safe
+----------------------------
+
+NumPy provides a minimal CPU-only version of LAPACK named ``lapack_lite`` that can
```
Comment (rgommers) [inline doc/release/upcoming_changes/26750.improvement.rst:4]: minor: everything in numpy is CPU-only, so this looks a slightly off. I guess you meant without SIMD or raw assembly. I'm not sure what to replace it with though - maybe "minimal and lower-performance"?

File: doc/release/upcoming_changes/26750.improvement.rst [hunk 6]
```diff
@@ -0,0 +1,12 @@
+`lapack_lite` is thread safe
```
Comment (rgommers) [inline doc/release/upcoming_changes/26750.improvement.rst:1]: perhaps insert "now" here?

**Review by rgommers — COMMENTED** [review]
Thanks Nathan. One more little bug in one of the last commits, plus two tiny suggestions for the release note

File: doc/release/upcoming_changes/26750.improvement.rst (hunk 5 above)
Comment (seberg) [reply doc/release/upcoming_changes/26750.improvement.rst]: how about "slow fallback version".

Comment (seberg) [reply doc/release/upcoming_changes/26750.improvement.rst]: woops, missed the outdated somehow, that is what it is already :).

**Review by rgommers — APPROVED** [review]
LGTM now, in it goes. Thanks Nathan! And thanks for the review Sebastian.

**Pull request merged**